### PR TITLE
PROGRAMMES-5756 Programme Footer

### PR DESCRIPTION
**Pull request by mattgreenham**
https://jira.dev.bbc.co.uk/browse/PROGRAMMES-5756
https://jira.dev.bbc.co.uk/browse/PROGRAMMES-5757

### CODE REVIEW

File: app/Resources/js/lazyload.js [hunk 1]
```diff
@@ -1,74 +1,136 @@
...
+            data_outer_div: 'data-lazyload-outerdiv'
         },
-        setOptions: function (options) {
+        setOptions : function (options) {
             this.options = $.extend({}, this.options, options);
         },
-        init: function () {
-            var _this = this;
+        init : function () {
+            var _this = this, timer;
```
Comment (votemike) [inline app/Resources/js/lazyload.js]: What's this witchcraft?

File: app/config/routing_3rd_party.yml [hunk 2]
```diff
@@ -74,3 +80,8 @@ programme_broadcasts:
     defaults: { _controller: FrameworkBundle:Redirect:redirect, route: '' }
     requirements: { pid: '[0-9b-df-hj-np-tv-z]{8,15}' }
     schemes: [http]
+
+network:
+    path: /{networkUrlKey}
```
Comment (votemike) [inline app/config/routing_3rd_party.yml]: Are we having this route?

File: src/DsAmen/Organism/Footer/FooterPresenter.php [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+use BBC\ProgrammesPagesService\Domain\Entity\Network;
+use BBC\ProgrammesPagesService\Domain\Entity\Programme;
+
+class FooterPresenter extends Presenter
+{
+    /** @var  Programme */
+    private $programme;
+
+    /** @var Network */
+    protected $network;
```
Comment (votemike) [inline src/DsAmen/Organism/Footer/FooterPresenter.php]: protected things above private ones please

File: src/DsAmen/Organism/Footer/FooterPresenter.php [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    {
+        parent::__construct($options);
+
+        $this->programme = $programme;
+        $this->network = $programme->getNetwork();
+    }
+
+    public function hasNetwork()
+    {
+        if ($this->network) {
```
Comment (votemike) [inline src/DsAmen/Organism/Footer/FooterPresenter.php]: Or just `return (bool) $this->network`?

File: src/DsAmen/Organism/Footer/FooterPresenter.php [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+    public function __construct(Programme $programme, array $options = [])
+    {
+        parent::__construct($options);
+
+        $this->programme = $programme;
+        $this->network = $programme->getNetwork();
+    }
+
+    public function hasNetwork()
```
Comment (votemike) [inline src/DsAmen/Organism/Footer/FooterPresenter.php]: Return types for everything?

**Comment by programmes-developers** [comment]
Can one of the admins verify this patch?

File: src/DsAmen/Organism/Footer/FooterPresenter.php (hunk 3 above)
Comment (BPScott) [reply src/DsAmen/Organism/Footer/FooterPresenter.php]: This class isn't going to be inherited from, so this should be private anyway.

File: app/config/routing_3rd_party.yml [hunk 6]
```diff
@@ -74,3 +80,8 @@ programme_broadcasts:
...
+    defaults: { _controller: FrameworkBundle:Redirect:redirect, route: '' }
+    requirements: { networkUrlKey: '[a-zA-Z]{2,35}' }
```
Comment (BPScott) [inline app/config/routing_3rd_party.yml]: urlkeys can have numbers in them too, check the format in the other routes

File: app/Resources/views/find_by_pid/tlec.html.twig [hunk 7]
```diff
@@ -80,3 +80,5 @@
 
 {# TODO Related links block #}
 {% endblock %}
+
+{% block footer %}{{ ds_amen('footer', programme) }}{% endblock %}
```
Comment (BPScott) [inline app/Resources/views/find_by_pid/tlec.html.twig]: Having a separate block seems unneeded, we could stick the {{dsamen()}} call at the bottom of the body block

File: gulpfile.js [hunk 8]
```diff
@@ -37,14 +37,15 @@ gulp.task('js', ['js:clean'], function () {
...
+            "jquery.appear": "../../../node_modules/jquery.appear/jquery.appear",
             "lazysizes": "../../../node_modules/lazysizes/lazysizes-umd",
             "eqjs": "../../../node_modules/eq.js/build/eq",
             "comscorews" : "../../../vendor/bbc-rmp/comscore/js-modules/comscorews",
             "rmpcomscore/base" : "../../../vendor/bbc-rmp/comscore/js-modules/base",
             "orb/cookies": "empty:"
         },
         // @TODO do not optimise on sandbox
-        //"optimize": "none",
+        "optimize": "none",
```
Comment (BPScott) [inline gulpfile.js]: Put this back

File: src/DsAmen/Organism/Footer/FooterPresenter.php (hunk 3 above)
Comment (BPScott) [reply src/DsAmen/Organism/Footer/FooterPresenter.php]: This class isn't being inherited from, so should be private.

File: src/DsAmen/Organism/Footer/footer.html.twig [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+                <div class="grid__item 1/3@gel4 1/4@gel3b 1/2@gel3">
+                    <div class="islet--horizontal footer__programmes footer__service-island">
+                        <h2 class="gel-double-pica-bold">{{ tr('similar_programmes') }}</h2>
+                        {%- if footer.getProgrammeGenreHierarchy() -%}
+                            <div class="islet--vertical">
+                                <h3 class="gel-brevier-bold text--shout">{{ tr('by_genre') }}:</h3>
+                                <ul class="list-raw list-unstyled text--shout gel-brevier-bold">
+                                    {%- for genreTree in footer.getProgrammeGenreHierarchy() -%}
+                                        <li>
+                                            {%- for genre in genreTree|reverse -%}
```
Comment (BPScott) [inline src/DsAmen/Organism/Footer/footer.html.twig]: If you make footer.getGenres() return a simple list, then this line can be `genre.getAncestry|reverse` which saves you having to loop through that array twice

File: app/Resources/js/lazyload.js [hunk 10]
```diff
@@ -1,4 +1,4 @@
-define(['jquery-1.9'], function ($) {
+define(['jquery-1.9', 'jquery.appear'], function ($) {
```
Comment (BPScott) [inline app/Resources/js/lazyload.js]: The original version of this didn't need appear, as it's functionality was pretty small, is there a way where we can not reintroduce that dependency?

File: src/DsAmen/Organism/Footer/FooterPresenter.php [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        return $this->network ? $this->network->getUrlKey() : "";
+    }
+
+    public function getNetworkName(): string
+    {
+        return $this->network ? $this->network->getName() : "";
+    }
+
+    public function getNavigationLinks(): array
```
Comment (votemike) [inline src/DsAmen/Organism/Footer/FooterPresenter.php]: Would you be able to have a look at this: https://gist.github.com/votemike/04d7f5c2859a6999a5aa3915ea40c982 Ask IRL if you don't understand

Comment (mattgreenham) [reply src/DsAmen/Organism/Footer/FooterPresenter.php]: Fixed

**Pull request merged**